aws_lambda_builders/workflows/java_gradle/DESIGN.md

```diff
@@ -0,0 +1,117 @@
+# Java - Gradle Lambda Builder
+
+## Scope
+
+This package enables the creation of a Lambda deployment package for Java
+projects managed using the Gradle build tool.
+
+For Java projects, the most popular way to create a distribution package for
+Java based Lambdas is to create an "uber" or "fat" JAR. This is a single JAR
+file that contains both the customers' classes and resources, as well as all the
+classes and resources extracted from their dependency JAR's. However, this can
+cause files that have the same path in two different JAR's to collide within the
+uber JAR.
+
+Another solution is to create a distribution ZIP containing the customer's
+classes and resources and include their dependency JARs under a `lib` directory.
+This keeps the customers' classes and resources separate from their
+dependencies' to avoid any file collisions. However, this incurs some overhead
+as the ZIP must be unpacked before the code can run.
+
+To avoid the problem of colliding files, we will choose the second option and
+create distribution ZIP's by default. We can further provide the `--uber-jar`
+option for customers to force the creation of an uber JAR if necessary. This
+distribution will only contain the libraries needed at runtime (i.e. the runtime
+classpath libraries).
+
+Finally, a `--use-container` option will be provided to address potential issues
+with environment-dependent build scripts (see Challenges).
+
+## Challenges
+
+Java bytecode can only run on the same or newer version of the JVM for which
+it was compiled for. For example Java 8 bytecode can run a JVM that is at
+least version 8, but bytecode targetting Java 9 cannot run on a Java 8 VM.
+This is further complicated by the fact that a newer JDK can generate code to
+be run on an older VM if configured using the `targetCompatibility` and
+`sourceCompatibility` properties of the Java plugin. Therefore, it is not
+sufficient to check the version of the local JDK, nor is it possible to check
+the value set for `targetCompatibility` or `sourceCompatibility` since it can
```

Comment:
with #55 , you can provide your own validator to check for this before the workflow begins to run.

Reply:
Can you elaborate on which part #55 can help with? `targetCompatibility` and `sourceCompatibility` would be set at build time within `build.gradle`.

Reply:
Ah, I just meant it as a check on the JDK version that was found when building for the java8 runtime.